libmamba/src/solv-cpp/repo.cpp

```diff
@@ -0,0 +1,278 @@
+            ss << "Unable to write file '" << solv_file << '\'';
+            throw std::runtime_error(ss.str());
+        }
+    }
+
+    /***********************************
+     *  Implementation of ObjRepoView  *
+     ***********************************/
+
+    ObjRepoView::ObjRepoView(::Repo* repo) noexcept
+        : ObjRepoViewConst{ repo }
+    {
+    }
+
+    auto ObjRepoView::raw() const -> ::Repo*
+    {
+        // Safe because we were passed a ``Repo*`` at construction time.
+        return const_cast<::Repo*>(ObjRepoViewConst::raw());
+    }
+
+    void ObjRepoView::clear(bool reuse_ids) const
+    {
+        ::repo_empty(raw(), static_cast<int>(reuse_ids));
+    }
+
+    void ObjRepoView::read(std::filesystem::path solv_file) const
+    {
+        auto* fp = std::fopen(solv_file.string().c_str(), "rb");
+        if (!fp)
+        {
+            throw std::system_error(errno, std::generic_category());
+        }
+        const auto read_res = ::repo_add_solv(raw(), fp, 0);
+        const auto close_res = std::fclose(fp);
+        if (read_res != 0)
+        {
+            // TODO(C++20) fmt::format
+            auto ss = std::stringstream();
+            ss << "Unable to read repo solv file '" << name() << '\'';
+            throw std::runtime_error(ss.str());
+        }
+        if (close_res != 0)
+        {
+            // TODO(C++20) fmt::format
+            auto ss = std::stringstream();
+            ss << "Unable to read from file '" << solv_file << '\'';
+            throw std::runtime_error(ss.str());
+        }
+    }
+
+    auto ObjRepoView::add_solvable() const -> std::pair<SolvableId, ObjSolvableView>
+    {
+        const SolvableId id = ::repo_add_solvable(raw());
+        return {
+            id,
+            get_solvable(id).value()  // Safe because we just added the solvable
+        };
+    }
+
+    auto ObjRepoView::get_solvable(SolvableId id) const -> std::optional<ObjSolvableView>
+    {
+        if (::Solvable* s = get_solvable_ptr(raw(), id); s != nullptr)
+        {
+            return { ObjSolvableView{ s } };
+        }
+        return std::nullopt;
+    }
+
+    auto ObjRepoView::remove_solvable(SolvableId id, bool reuse_id) const -> bool
+    {
+        if (has_solvable(id))
+        {
+            ::repo_free_solvable(raw(), id, reuse_id);
+            return true;
+        }
+        return false;
+    }
+
+    void ObjRepoView::internalize()
+    {
+        ::repo_internalize(raw());
+    }
+
+    /****************************************
+     *  Implementation of getter & setters  *
+     ****************************************/
+
+    auto ObjRepoViewConst::name() const -> std::string_view
+    {
+        return ::repo_name(raw());
+    }
+
+    namespace
+    {
+        auto ptr_to_strview(const char* ptr) -> std::string_view
+        {
+            static constexpr std::string_view null = "<NULL>";
+            if ((ptr == nullptr) || (ptr == null))
+            {
+                return "";
```

Comment:
Is this equivalent to `return {}`? Should it be?

Reply:
I think it is not exactly the same in term of `.data` and. I have no preference between the two.